File: algorithms/2_medium/C/01170_compare_strings-by_frequency_of_the_smallest_character/prefix.c
#define MAX_ALLOWED_FREQ 10

static size_t calc_freqs_smallest_char(const char* s)
{
    assert(s != NULL);
    char curr_ch = 'z';
    size_t freq = 0;

    for (size_t i = 0; s[i] != '\0'; i += 1) {
        if (s[i] < curr_ch) {
            curr_ch = s[i];
            freq = 1;
        } else if (s[i] == curr_ch) {
            freq += 1;
        }
    }

    return freq;
}

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* numSmallerByFrequency(char** queries, const int queries_len,
                           char** words, const int words_len, int* return_len)
{
    assert(queries_len > 0 && words_len > 0 && return_len != NULL && queries != NULL && words != NULL);
    assert(words_len <= 2000 && queries_len <= 2000);


    int16_t prefix[MAX_ALLOWED_FREQ + 1] = { 0 };
    for (int i = 0; i < words_len; i += 1) {
        const size_t freq = calc_freqs_smallest_char(words[i]);
        assert(freq <= MAX_ALLOWED_FREQ);
        prefix[freq] += 1;
    }
    for (size_t i = 1; i <= MAX_ALLOWED_FREQ; i += 1) {
        prefix[i] += prefix[i - 1];
    }

    int* res = malloc((size_t)queries_len * sizeof(*res));
    assert(res != NULL);
    *return_len = queries_len;

    for (int i = 0; i < queries_len; i += 1) {
        const size_t freq = calc_freqs_smallest_char(queries[i]);
        if (freq >= MAX_ALLOWED_FREQ) {
            res[i] = 0;
        } else {
            res[i] = prefix[MAX_ALLOWED_FREQ] - prefix[freq];
        }
    }

    return res;
}

File: algorithms/2_medium/C/01170_compare_strings-by_frequency_of_the_smallest_character/prefix.rs
impl Solution {
    const MAX_ALLOWED_FREQ: usize = 10;

    #[inline(always)]
    fn calc_freqs_smallest_char(bytes: &[u8]) -> usize {
        let mut curr_byte = b'z';
        let mut freq = 0_usize;

        for &b in bytes {
            assert!(b >= b'a' && b <= b'z');
            if b < curr_byte {
                curr_byte = b;
                freq = 1;
            } else if b == curr_byte {
                freq += 1;
            }
        }

        freq
    }

    pub fn num_smaller_by_frequency(queries: Vec<String>, words: Vec<String>) -> Vec<i32> {
        assert!(words.len() <= 2000 && queries.len() <= 2000);
        let mut prefix = [0_i16; Self::MAX_ALLOWED_FREQ + 1];

        for w in words {
            let freq = Self::calc_freqs_smallest_char(w.as_bytes());
            assert!(freq <= Self::MAX_ALLOWED_FREQ);
            prefix[freq] += 1;
        }
        for i in 1..=Self::MAX_ALLOWED_FREQ {
            prefix[i] += prefix[i - 1];
        }

        let mut res: Vec<i32> = Vec::with_capacity(queries.len());
        unsafe { res.set_len(queries.len()); }
        for (i, q) in queries.into_iter().enumerate() {
            let freq = Self::calc_freqs_smallest_char(q.as_bytes());
            if freq >= Self::MAX_ALLOWED_FREQ {
                res[i] = 0;
            } else {
                res[i] = (prefix[Self::MAX_ALLOWED_FREQ] - prefix[freq]) as i32;
            }
        }

        res
    }
}
